1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,279 @@
+- Changes will be required for `.ForEach{}` and `.Where{}` magic method implementations, which have some custom handling for named blocks that will need to be updated to account for `cleanup {}`.
+- Improvements to `PipelineStopper` to be able to properly recognise critical regions of code (like `cleanup {}`) and prevent pipeline stops from being registered until that region of code is exited.
+  - Ctrl+C will still be registered, but the pipeline stop will not begin until that critical code region is exited.
+
+#### Error States
+
+Terminating errors can be thrown from a `cleanup {}` block, which will skip the invocation of the rest of that `cleanup {}` block, without preventing other commands' `cleanup {}` blocks from being executed.
+
+#### Ctrl+C Behaviour
+
+While scripts usually terminate automatically after the StopProcessing signal is received, an exception will be made during the `cleanup {}` sequences, and the `PipelineStopper.IsStopping` flag will be temporarily cleared to permit script execution for the purposes of `cleanup {}`.
+This should allow us to ensure the block executes correctly during the pipeline stopping sequence.
+
+`cleanup {}` code should run regardless of `Ctrl+C` being pressed, whether before or during its operation, so that authors can ensure critical code can be run even when the user cancels normal processing.
+
+#### Finalizer
+
+As was raised during discussion with the committee, we may need to handle the edge case that a script cmdlet is not properly disposed by the time the finalizer is called.
+This _should_ be an impossible state to reach since the PowerShell engine typically has full control over the lifecycle of a command, but since the implementation proposed includes having `PSScriptCmdlet` implement `IDisposable` directly, it's possible its `cleanup {}` block could be invoked on the finalizer thread and subsequently throw an error as there would not be a PowerShell runspace available to execute the code.
+
+Similar to `InvokeWithPipe()` overloads, we can check if the current thread context matches the original context of the command.
+If we find that the invocation is happening on the wrong thread, we can raise an event in the original context to have it executed there instead.
+
+#### `Cmdlet` / `PSCmdlet`
+
+Commands inheriting from `Cmdlet` or `PSCmdlet` can already implement a version of this behaviour.
+To do so, such commands need also implement `IDisposable` and the `Dispose()` method.
+This RFC would simply change when the dispose method is called.
+Currently, the method is called during pipeline disposal, after **all** commands have completed their `EndProcessing()` steps.
+As part of the changes proposed in this RFC, we would ensure each command's `cleanup {}` or `Dispose()` routines are called immediately after _each_ cmdlet's `EndProcessing()` execution completes during normal pipeline operation.
```

Comment:
i'm nervous about changing the ordering here with regard to when `Dispose` is called. What behavioral change do you think will happen because of this change, and if none, is there a way that you can validate that.

Reply:
There shouldn't be a _lot_ of visible changes here, but there would be some. Essentially the change here would simple reorder the EndProcessing/Dispose steps in a given pipeline from `EndA -> EndB -> EndC -> DisposeA -> DisposeB -> DisposeC` to instead being `EndA -> DisposeA -> EndB -> DisposeB -> EndC -> DisposeC`

To an individual command, there should not be a significant difference. It's still getting disposed after it completes EndProcessing, just with a bit less delay than before.

For a practical example, let's take the example of a cmdlet at the start of a pipeline that takes a lock on a file to read data in during `BeginProcessing()` and doesn't release it until `Dispose()`. Suppose also that it writes all of its output in `EndProcessing()`. Then you want to pipe that to another command that wants to also take a write lock on that same file, but doesn't need it until `EndProcessing()`.

Contrived example? Maybe!

Currently, this circumstance will fail completely, because the first command's lock is still alive by the time the second command wants to take a lock. If we move disposal/cleanup to just after each command's `EndProcessing()` instead, then everything just works.

Hugely useful scenario? Maybe, maybe not. It's kind of an edge case where it will be super useful, but at least on paper I think it makes more sense to dispose resources and do cleanup when the command has done its job, rather than when the pipeline as a whole is completed. 

If we suppose that all the commands in a given pipeline have some `cleanup{}` or `Dispose()` processing going on, doing it after each command will "feel" better overall to users, I think; it'll spread out any impact of running cleanup/disposal across the run of the pipeline, rather than effectively queueing it up until the end of the pipe.

---

Anyway, that's my reasoning.

As for do I think there'll be a behavioural change... no, in most cases there won't be. Can I validate that.... maybe? Not sure. I can't really think of a way apart from simply demonstrating that commands in general still work just fine, even when they have implement `IDisposable` or a `cleanup{}` block.

I'd venture to guess that if it were a problem, we'd see some odd behaviour fairly quickly from cmdlets that utilise the IDisposable pattern currently? Outside of those, no commands presently written could realistically be affected, though.

Reply:
(sorry for the long text, didn't realize I'd gone on so long 😬 )